### DISCUSSION TITLE
ALPN failed: what does this mean?

### DISCUSSION THREAD

**Discussion by jpsollie** [openssl/openssl#25606] [Q&A] [answered]
I'm trying to move my own bind server to use DoT instead of regular TCP / UDP transport.  I picked a few having DoT advertised, but almost every connection attempt ends with error "ALPN for DoT failed".
This seems to be a problem with openssl tls negotiation: it shows up with a few other applications when I google it.
Unfortunately, googling it couldn't provide me a useful answer of how to deal with these errors.
Does anyone here have a clue of where to search in the first place?  why does it fail?

**Comment by nhorman** [comment]
how are you setting the alpn in your code?  Can you provide a code snippet?

**Comment by jpsollie** [comment]
A small C/P from my config file:
```
tls ns28 {
  cert-file "/data/cache/bind/keys/domain.crt";
  key-file "/data/cache/bind/keys/domain.pass.key";
  protocols { TLSv1.2; TLSv1.3; };
  ciphers "HIGH:MEDIUM:LOW";
  prefer-server-ciphers yes;
  session-tickets no;
};
```

**Reply by nhorman** [reply]
I don't see how that specifies your alpn

**Reply by jpsollie** [reply]
then you already know a lot more than I do.  So, can I tcpdump such a connection attempt and get some useful info about the ALPN stuff? what it tries and why it fails?

**Reply by nhorman** [reply]
Sure, a tcpdump never hurts, but it sounds like your bind server is not something you've written, just something you're using?  It may just be that the bind server doesn't support the dot alpn.  Can you point me to what you're using?

**Reply by mattcaswell** [reply, marked as answer]
This really sounds more like a question to submit to some bind forum. We're not bind configuration or DNS experts, so we can't help you there. However, I can tell you at a TLS level what this means.

ALPN is Application Layer Protocol Negotiation. This is a TLS extension that enables you to negotiate what application protocol is going to be used over the eventual TLS connection that gets established.

The list of well defined ALPN values is here:
https://www.iana.org/assignments/tls-extensiontype-values/tls-extensiontype-values.xhtml#alpn-protocol-ids

For DoT the ALPN value is "dot". The TLS client sends this ALPN value to the server, and the server is expected to confirm that it is able to support this protocol.

What appears to be happening in your case is that the "dot" ALPN value is being sent to the server, but the server is not confirming it - hence "ALPN for DoT failed".

As to why, this is probably more a bind/DNS configuration issue.

**Reply by jpsollie** [reply]
thank you for the explanation

**Reply by mattcaswell** [reply]
A further thought on this: this kind of error can happen if you have mismatched protocols. For example if you try to talk DNS to a web server, then you might get this problem. Just a thought.

**Reply by jpsollie** [reply]
I do not think that's the issue here: if I specify the remote client is at port 853, it's unlikely to answer as "http" ...
simply to investigate firther what's exactly happening, is there something I should see on tcpdump what may be important? eg: are there flags in the tls handshake specifying "alpn is required"? and which alpns the servers wants to support? it may very well be the key to fixing the issue

**Reply by nhorman** [reply]
Take a look at the client and server hello messages, in the transport params, you'll see the exchanged offered alpn values from the client, and the selected alpn the server responds with.

**Reply by mattcaswell** [reply]
I can demonstrate a client and server looking to negotiate the "dot" ALPN using OpenSSL's s_client and s_server apps. Starting s_server like this (using a suitable cert and key):

````
openssl s_server -cert cert.pem -key key.pem -alpn dot
````

I can connect to it using s_client like this:

````
openssl s_client -alpn dot -trace -tls1_2
````
Which gives me this output:

````
CONNECTED(00000003)
Sent Record
Header:
  Version = TLS 1.0 (0x301)
  Content Type = Handshake (22)
  Length = 211
    ClientHello, Length=207
      client_version=0x303 (TLS 1.2)
      Random:
        gmt_unix_time=0x42498A09
        random_bytes (len=28): DD6B79F52594480FDB10BEC61E796FA59A6CB7DF024ABE948E14D885
      session_id (len=0): 
      cipher_suites (len=56)
        {0xC0, 0x2C} TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
        {0xC0, 0x30} TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
        {0x00, 0x9F} TLS_DHE_RSA_WITH_AES_256_GCM_SHA384
        {0xCC, 0xA9} TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256
        {0xCC, 0xA8} TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256
        {0xCC, 0xAA} TLS_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256
        {0xC0, 0x2B} TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
        {0xC0, 0x2F} TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
        {0x00, 0x9E} TLS_DHE_RSA_WITH_AES_128_GCM_SHA256
        {0xC0, 0x24} TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384
        {0xC0, 0x28} TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384
        {0x00, 0x6B} TLS_DHE_RSA_WITH_AES_256_CBC_SHA256
        {0xC0, 0x23} TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256
        {0xC0, 0x27} TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256
        {0x00, 0x67} TLS_DHE_RSA_WITH_AES_128_CBC_SHA256
        {0xC0, 0x0A} TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA
        {0xC0, 0x14} TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA
        {0x00, 0x39} TLS_DHE_RSA_WITH_AES_256_CBC_SHA
        {0xC0, 0x09} TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA
        {0xC0, 0x13} TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA
        {0x00, 0x33} TLS_DHE_RSA_WITH_AES_128_CBC_SHA
        {0x00, 0x9D} TLS_RSA_WITH_AES_256_GCM_SHA384
        {0x00, 0x9C} TLS_RSA_WITH_AES_128_GCM_SHA256
        {0x00, 0x3D} TLS_RSA_WITH_AES_256_CBC_SHA256
        {0x00, 0x3C} TLS_RSA_WITH_AES_128_CBC_SHA256
        {0x00, 0x35} TLS_RSA_WITH_AES_256_CBC_SHA
        {0x00, 0x2F} TLS_RSA_WITH_AES_128_CBC_SHA
        {0x00, 0xFF} TLS_EMPTY_RENEGOTIATION_INFO_SCSV
      compression_methods (len=1)
        No Compression (0x00)
      extensions, length = 110
        extension_type=server_name(0), length=14
          0000 - 00 0c 00 00 09 6c 6f 63-61 6c 68 6f 73 74      .....localhost
        extension_type=ec_point_formats(11), length=4
          uncompressed (0)
          ansiX962_compressed_prime (1)
          ansiX962_compressed_char2 (2)
        extension_type=supported_groups(10), length=12
          ecdh_x25519 (29)
          secp256r1 (P-256) (23)
          ecdh_x448 (30)
          secp521r1 (P-521) (25)
          secp384r1 (P-384) (24)
        extension_type=session_ticket(35), length=0
        extension_type=application_layer_protocol_negotiation(16), length=6
          dot
        extension_type=encrypt_then_mac(22), length=0
        extension_type=extended_master_secret(23), length=0
        extension_type=signature_algorithms(13), length=42
          ecdsa_secp256r1_sha256 (0x0403)
          ecdsa_secp384r1_sha384 (0x0503)
          ecdsa_secp521r1_sha512 (0x0603)
          ed25519 (0x0807)
          ed448 (0x0808)
          rsa_pss_pss_sha256 (0x0809)
          rsa_pss_pss_sha384 (0x080a)
          rsa_pss_pss_sha512 (0x080b)
          rsa_pss_rsae_sha256 (0x0804)
          rsa_pss_rsae_sha384 (0x0805)
          rsa_pss_rsae_sha512 (0x0806)
          rsa_pkcs1_sha256 (0x0401)
          rsa_pkcs1_sha384 (0x0501)
          rsa_pkcs1_sha512 (0x0601)
          ecdsa_sha224 (0x0303)
          rsa_pkcs1_sha224 (0x0301)
          dsa_sha224 (0x0302)
          dsa_sha256 (0x0402)
          dsa_sha384 (0x0502)
          dsa_sha512 (0x0602)

Received Record
Header:
  Version = TLS 1.2 (0x303)
  Content Type = Handshake (22)
  Length = 75
    ServerHello, Length=71
      server_version=0x303 (TLS 1.2)
      Random:
        gmt_unix_time=0x9748C440
        random_bytes (len=28): C3E70126758AC57662219EC13E87F97681706377444F574E47524401
      session_id (len=0): 
      cipher_suite {0xC0, 0x30} TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
      compression_method: No Compression (0x00)
      extensions, length = 31
        extension_type=renegotiate(65281), length=1
            <EMPTY>
        extension_type=ec_point_formats(11), length=4
          uncompressed (0)
          ansiX962_compressed_prime (1)
          ansiX962_compressed_char2 (2)
        extension_type=session_ticket(35), length=0
        extension_type=application_layer_protocol_negotiation(16), length=6
          dot
        extension_type=extended_master_secret(23), length=0
...snip...
````

In the above you can see the client sending a ClientHello message, with the application_layer_protocol_negotiation(16) extension set to "dot". You can then see the server responding with its ServerHello, echoing the same application_layer_protocol_negotiation(16) "dot" value.

This is a TLSv1.2 connection which is easier to handle because both the client and server hellos are in plaintext unencrypted. You should be able to see this in a tcpdump capture. Unfortunately, the protocol doesn't tell you what ALPN values the server supports - only if the ALPN requested by the client was successfully chosen. Also, if you end up with a TLSv1.3 connection (which is actually probably more likely) then unfortunately the ServerHello message does not have an alpn extension and instead it is sent in the EncryptedExtensions message (which as the name suggests is not sent in plaintext). So it gets more difficult to trace the output with tcpdump in this case.

**Reply by jpsollie** [reply]
hmm ... I'm trying your commands on all DoT servers I registered, but the only output I see at the end is:
```
New, TLSv1.2, Cipher is ECDHE-ECDSA-CHACHA20-POLY1305
Server public key is 256 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-ECDSA-CHACHA20-POLY1305
```
is there some way to force negotiation?

**Reply by nhorman** [reply]
That depends on the server implementation.  As you can see from @mattcaswell example, prior to the output you posted, the alpn is negotiated in the client and server hello records.  the client hello sends, as part of its extension list, this extension:
```
extension_type=application_layer_protocol_negotiation(16), length=6
          dot
```
indicating that the client would like to use the alpn of dot

the server in response, should send back a similar extention, which in his example looks like this:
```
extension_type=application_layer_protocol_negotiation(16), length=6
          dot
```
because each only sends a single alpn value, and they match, thats the alpn that is negotiated/agreed upon.

alternatively the server may send back a different set of alpns.  if there is no match, then no alpn is negotiated

So to answer your direct question, there is  no way to "force" negotiation.  What you have to do instead (this is subtly different), is force the server to offer dot as an alpn protocol during the handshake.  How to do that is a matter of understanding if/how your bind server offers a mechanism to do that, and thats something we can't help with, you'll have to contact your bind server vendor/project to figure that out

**Reply by jpsollie** [reply]
all right.  Thank you for all explanation and help diagnosing the issue!!

**Reply by mattcaswell** [reply]
Just another data point on this:

I tried connecting to one of the publicly listed DoT servers [here](https://dnsprivacy.org/public_resolvers/), like this:

````
openssl s_client -alpn dot -connect dns.quad9.net:853 -trace
````

And saw a successful ALPN negotiation:

````
New, TLSv1.3, Cipher is TLS_AES_256_GCM_SHA384
Protocol: TLSv1.3
Server public key is 256 bit
This TLS version forbids renegotiation.
Compression: NONE
Expansion: NONE
ALPN protocol: dot
````
